src/elexclarity/convert.py

```diff
@@ -1,7 +1,132 @@
+            # more validation, for one precinct races
+            precincts = i["Precinct"]
+            if type(precincts) != list:
+                precincts = [precincts]
+            
+            for subunit in precincts:
+                subunits[subunit["name"]] += int(subunit["votes"])
+        
+        return {
+            "key": key,
+            "name": name,
+            "id": slug,
+            "subunits": subunits,
+        }
+
+    def aggregate_subunits_from_choices(self, choices, fips):
+        """
+        Takes a list of `Choice` objects from Clarity and aggregates/transforms
+        them into a the format our data importer expects.
+        """
+        processed_choices = [self.get_subunit_totals_from_choice(i) for i in choices]   
+
+        # Get a flat, unique, list of our subunit names
+        subunits = [i["subunits"].keys() for i in processed_choices]
+        subunits = set([self.get_id_for_precinct(fips, i) for l in subunits for i in l])
+        
+        # Get a list of our candidates
+        candidates = [i['id'] for i in processed_choices]
+        
+        agg = {i: {
+            "id": i,
+            "counts": defaultdict(lambda: 0)
+        } for i in subunits}
+
+        for choice in processed_choices:
+            candidate = choice["id"]
+            for subunit_name, vote_count in choice["subunits"].items():
+                key = self.get_id_for_precinct(fips, subunit_name)
+                agg[key]["counts"][candidate] += vote_count
+        
+        return agg
+    
+    def get_total_votes_from_choices(self, choices):
+        agg = {}
+        
+        for choice in choices:
+            name = choice.get("text")
+            key = choice.get("key")
+            slug = self.get_id_for_choice(key, name)
+            agg[slug] = int(choice["totalVotes"])
+        
+        return agg
+
+    def transform_contest(self, contest, fips=""):
+        precincts_reported = int(contest.get("precinctsReported"))
+        precincts_reporting = int(contest.get("precinctsReporting"))
+        
+        # some light validation on the choices to make sure we get a list
+        choices = contest["Choice"]
+        if type(choices) != list:
+            choices = [choices]        
+        
+        return {
+            "source": "clarity",
+            "name": contest.get("text"),
+            "precinctsReportingPct": precincts_reported/precincts_reporting,
+            "subunits": self.aggregate_subunits_from_choices(choices, fips=fips),
+            "counts": self.get_total_votes_from_choices(choices)
+        }
+
+    def transform_result_object(self, result):
+        county = result["Region"]
+        fips = self.county_lookup.get(county)
+        contests = [self.transform_contest(i, fips=fips) for i in result["Contest"]]
+        return {i["name"]: i for i in contests}
+
+    def convert(self, data, statepostal="", outputType="results", style="default", resultsBy=None, **kwargs):
```

Comment:
I know I am the queen of forcing everything into OOP, but the [existing setup](https://github.com/WPMedia/elex-data-importer/blob/develop/elexdata/importer.py#L3) is expecting not an instance method

Reply:
I would keep this individual `convert` method here (but detach it from the class), name this class `ClarityXMLPrecinctConverter`, and move it elsewhere (i.e. use [the elex-edison/elex-ap structure](https://github.com/WPMedia/elex-edison/blob/develop/src/elexedison/convert.py#L78) )